javalib/src/main/scala/java/nio/file/attribute/PosixFileAttributeViewImpl.scala

```diff
@@ -86,8 +87,34 @@ final class PosixFileAttributeViewImpl(path: Path, options: Array[LinkOption])
 
   override def readAttributes(): BasicFileAttributes = attributes
 
+  private class PosixFileKey(deviceId: stat.dev_t, inodeNumber: stat.ino_t) {
+
+    override def equals(that: Any): Boolean = that match {
+      case n if that == null => false
+      case that: Object      => this.hashCode() == that.hashCode()
+      case _                 => false
+    }
```

Comment:
We should check for equality of each of the fields.
```suggestion
    override def equals(that: Any): Boolean = that match {
      case that: PosixFileKey      => this.deviceId == that.deviceId && this.inodeNumber == that.inodeNumber
      case _                 => false
    }
```

Reply:
Wojciech,

Thank you for the improvements.  

1) Doesn't your suggestion have a problem with `that` being null (NPE)?  Now I am not a big fan of 
    nulls being being passed around but it it part of the definition of `BasicFileAttributes#fileKey` that
    it can return null.

2) re: `case that: PosixFileKey`  PosixFileKey is a Scala Native implementation-only definition.
    Java `BasicFileAttributes#fileKey` returns `Object`. If IIUC, `that` will always be an `Object` 
    and never activate the `case that: PosixFileKey`.
      
    If the code under discussion is executing, then `this` must know it is a `PosixFileKey` extension
    of `Object`. Is there an idiomatic way to say "case that if classOf[that] == classOf[this]"?

    I know that `hasCode()` matches are inexact and much prefer the `==` suggested (to fast longword
    machine comparisons) if I can convince myself and the runtime that `that` is a `PosixFileKey`. 

Sorry to take time on this apparent molehill.  

I hope to use the fileKey for doing File System Loop detection in javalib  `Files.walk` and `Files.walkFileTree`, 
at least on non-Windows. Debugging those is hard enough. It will be neigh on impossible to get right
if I mess up this primitive.   I have to work with  set-in-stone Java definitions here and would use less difficult
definitions.

Reply:
if the `that` is null it would not be matched in ` case that: PosixFileKey ` -  `isInstanceOf` on null is safe it always returns false, so the pattern would never be matched. 

> that will always be an Object and never activate the case that: PosixFileKey. 

 `Object` is a compile-time type, that's the most specific type we get from the signatures. In the equals we check for runtime type and it would typically be `PosixFileKey`
 
> part of the definition of BasicFileAttributes#fileKey that it can return null. 

If `that` is null it's handled by the wildcard case which returns false. If we get to the point when we need to check if it's a null then `this` is already non-null - we cannot call method on null, so the result of equals should return false.
 
 > Is there an idiomatic way to say "case that if classOf[that] == classOf[this]"? 
 
It could be `case that if that.getClass == this.getClass`, but in case of `PosixFileKey` is literally just `case PosixFileKey => ` - there is no other case we're intrested about. The class only needs to check if some other object can equal this object. There is no universal equality here, and `a.equals(b)` might not always be the same as `b.equals(a)` depending on implementation of `equals` in A and B. 

Note: there is one more case the equals method should check first - it should start with `this eq that` to make a quick check for referential equality, before starting the value equality.

Reply:
Thank you for the patient and informative explanation.  

I will implement your suggestion later today.

Reply:
+1
> Thank you for the patient and informative explanation.